src/layers/SchematicLayer.ts

```diff
@@ -305,7 +323,46 @@ export class SchematicLayer<T extends SchematicData> extends PixiLayer<T> {
     }
 
     const { exaggerationFactor } = this.options as SchematicLayerOptions<T>;
-    const { holeSizes, casings, cements, completion, symbols, pAndA } = this.data;
+    const { holeSizes, casings, cements, completion, symbols, pAndA, perforations } = this.data;
+
+    const shouldStartAtHoleDiameter = perforations.filter((perf) =>
+      foldPerforationSubKind(
+        {
+          Perforation: () => true,
+          OpenHole: () => false,
+          OpenHoleGravelPack: () => true,
+          OpenHoleFracPack: () => false,
+          OpenHoleScreen: () => true,
+          CasedHoleFracturation: () => false,
+          CasedHoleGravelPack: () => false,
+          CasedHoleFracPack: () => false,
+        },
+        perf.subKind,
+      ),
+    );
+
+    const shouldStartAtCasingDiameter = perforations.filter(
+      (perf) =>
+        !foldPerforationSubKind(
+          {
+            Perforation: () => true,
+            OpenHole: () => false,
+            OpenHoleGravelPack: () => true,
+            OpenHoleFracPack: () => false,
+            OpenHoleScreen: () => true,
+            CasedHoleFracturation: () => false,
+            CasedHoleGravelPack: () => false,
+            CasedHoleFracPack: () => false,
+          },
+          perf.subKind,
+        ),
+    );
+
+    shouldStartAtHoleDiameter.forEach((perforation) => {
+      const perfShape = this.createPerforationShape(perforation, casings, holeSizes).map((p) => ({ ...p, diameter: p.diameter * 4 }));
+      const otherPerforations = perforations.filter((p) => p.id !== perforation.id);
+      this.drawComplexRope(perfShape, this.createPerforationTexture(perforation, perfShape, otherPerforations));
+    });
```

Comment:
I reckon this should render before holes?
Can we move these two cases into its own methods? 😄